language/move-binary-format/src/deserializer.rs

```diff
@@ -43,9 +43,21 @@ impl CompiledModule {
         binary: &[u8],
         max_binary_format_version: u32,
     ) -> BinaryLoaderResult<Self> {
-        let module = deserialize_compiled_module(binary, max_binary_format_version)?;
-        BoundsChecker::verify_module(&module)?;
-        Ok(module)
+        let prev_state = move_core_types::state::set_state(VMState::DESERIALIZER);
```

Comment:
Do you mind explain a bit why we need this global variable? It doesn't seem to me that we are using this anywhere? Is it just for external observability?

Reply:
This would be relevant for panic handlers that might want to change behavior depending on where the panic originates. 

> This PR also exposes an API to set and query the state of the Move VM via a shared thread-local. This can be useful for custom panic handlers that want to abort unless the code is executing inside a known panic handler (such as deserialization or verification).

Reply:
the point @chen-robert is making here is quite important and let me try to give it another spin as well (sorry if well understood and repeated once again).
If any system embedding the VM has a global and custom panic handler installed, that handler will be called before the `catch_unwind` here. If, say, that global handler exited the process (as for instance the custom handler installed by the blockchain process in the original Libra code) the `catch_unwind` here would never run.
The principle here (and the conversation we had) was that a global custom handler could check the thread local and see the state of it. And if inside a verification or deserialization state, do nothing - or something specific - knowing that the `catch_unwind` may be called.